mm2src/coins/z_coin.rs

```diff
@@ -376,6 +377,52 @@ impl ZCoin {
         }
     }
 
+    /// Gets spendable notes for transaction generation, ensuring previous transactions are confirmed/mined.
+    ///
+    /// # Flow
+    /// 1. Checks for any previous transaction with change that needs confirmation
+    /// 2. If found, waits up to 30 minutes for required confirmations (checking every 15 seconds)
+    /// 3. Once confirmed or if no previous tx, returns ordered list of spendable notes
+    async fn spendable_notes_required_for_tx(&self, required: &BigDecimal) -> MmResult<Vec<SpendableNote>, GenTxError> {
+        let prev = self.z_fields.previous_tx_with_change.lock().unwrap().take();
```

Comment:
I think this has an edge case bug on restart. If the user restarts the wallet and the previous tx with change is still unconfirmed, the spendable balance will be wrong and same issues can occur. There can also be more than one unconfirmed transaction with change outputs, i think we need to persist unconfirmed change across restarts somehow.

Reply:
I am looking how komodod wallet works with private txns:
it processes two cases:
- when a block comes 
- when a tx is created by the user.
For both cases it updates notes and nullifiers from my txns in the wallet.db.
Looks like, we do not do the second case. Why not?

Reply:
I don't think this solves this issue..

anyways I have properly reimplemented this feature with saving multiple change notes, cleanup method in swaps ops to do some additional work after swap is finished e.g clearing pending tx note logs e.t.c If this is endorsed I will proceed to implement persistent storage for it.

Reply:
As I can see, librustcash allows to store unconfirmed sent transactions in the db, for e.g. with decrypt_and_store_transaction(). Then transactions should be updated with block height, when confirmed.
And, get_balance() and select_spendable_notes() correctly take into account only confirmed notes.
So I guess, we would use the db to return also unspendable balance, to solve the issue @shamardy pointed out with the app restart.

(I also did tests with ZOMBIE chain:
z_listunspent and z_getbalance rpc by default return only confirmed notes and balance, but may return also unconfirmed if minconf set 0.
z_sendmany returns error if try to set minconf to 0 to spend from z-address, obviously, as we can't create nullifiers)

Reply:
> If the user restarts the wallet and the previous tx with change is still unconfirmed, the spendable balance will be wrong and same issues can occur.

I guess, in the current PR, if the user restarts the wallet while the tx with change was still unconfirmed, we would lose only the unspendable balance. (Also the spendable balance will be updated when the tx gets confirmed and we process its block)

Reply:
>> (I also did tests with ZOMBIE chain:
z_listunspent and z_getbalance rpc by default return only confirmed notes and balance, but may return also unconfirmed if minconf set 0.
z_sendmany returns error if try to set minconf to 0 to spend from z-address, obviously, as we can't create nullifiers)

Seems this is limited to only native mode.

Reply:
apparently, librustzcash doesn't store change output in db, instead it only discover and store it when scanning blocks that is after a any tx that created them is confirmed.

 I add a persistent storage to track these notes from kdf which fixes the issue aswell

Reply:
yes librustzcash does not discover the change when it stores unconfirmed notes.
However apparently, it stores notes it could decrypt in the received_notes table w/o block (https://github.com/KomodoPlatform/librustzcash/blob/4e030a0f44cc17f100bf5f019563be25c5b8755f/zcash_client_sqlite/src/lib.rs#L466). 
I thought, maybe we could just query those notes to show them as 'unconfirmed balance' like we now query the notes with the non-null block in the get_balance fn.

Reply:
https://github.com/KomodoPlatform/librustzcash/blob/4e030a0f44cc17f100bf5f019563be25c5b8755f/zcash_client_sqlite/src/lib.rs#L466 is still after a transaction has been confirmed since `nullifier` can't be null https://github.com/KomodoPlatform/librustzcash/blob/4e030a0f44cc17f100bf5f019563be25c5b8755f/zcash_client_sqlite/src/wallet/init.rs#L71. So `librustzcash` doesn't actually keep track of change notes until the parent transaction is confirmed which will then be stored in `received_notes`.

Reply:
Yes, thank you.
Looks like decrypt_and_store_transaction is not fully implemented yet

Reply:
FYI: since this [zcash librustzcash release](https://github.com/zcash/librustzcash/commit/67b84c25e06c0009b73d8c0afd91f28393cba4de) they started to track unconfirmed notes in the db and collect [change_pending_confirmation](https://github.com/zcash/librustzcash/blob/67b84c25e06c0009b73d8c0afd91f28393cba4de/zcash_client_backend/src/data_api.rs#L68C9-L68C36) amount.

Reply:
let's close this discussion.
(and open a new one, more related to the newest code)